app/src/main/java/unipd/se18/ocrcamera/ImageProcessing.java

```diff
@@ -0,0 +1,348 @@
+        List<MatOfPoint> contours = new ArrayList<>();
+        int mode = 0;
+        int method = 1;
+        Imgproc.findContours(dilatated, contours, new Mat(), mode, method);
+        //The third parameter contains additional information that is unused
+
+
+        /*
+            EXPERIMENTAL:
+            finds the text rectangle with the largest area
+            and saves it in "max_contour"
+         */
+        double maxArea = 0;
+        MatOfPoint max_contour = new MatOfPoint();
+        Iterator<MatOfPoint> iterator = contours.iterator();
+        while (iterator.hasNext()) {
+            MatOfPoint contour = iterator.next();
+            double area = Imgproc.contourArea(contour);
+            if (area > maxArea) {
+                maxArea = area;
+                max_contour = contour;
+            }
+        }
+
+        //Creates a rotated rectangle based on "max_contour"
+        RotatedRect rect = Imgproc.minAreaRect(new MatOfPoint2f(max_contour.toArray()));
+        return rect;
+    }
+
+
+    /**
+     * Crop the matrix with the given rectangle
+     * @param rectangle the part of the image you want to crop
+     * @param mat the matrix you want to crop
+     * @return a matrix that contains only the rectangle
+     * @author Thomas Porro(g1), Oscar Garrido (g1)
+     */
+    public Mat crop(RotatedRect rectangle, Mat mat) {
+
+        RotatedRect rect;
+
+        //Matrices we'll use
+        Mat rotationMat = new Mat();
+        Mat rotatedImg = new Mat();
+        Mat croppedImg = new Mat();
+
+        //Get angle and size from the bounding box
+        double angle = rectangle.angle;
+        Size rect_size = rectangle.size;
+
+        //Thanks to http://felix.abecassis.me/2011/10/opencv-rotation-deskewing/
+        if (rectangle.angle < -45.) {
+            angle += 90.0;
+            double width = rect_size.width;
+            double height = rect_size.height;
+            rect_size.width = height;
+            rect_size.height = width;
+        }
+
+        //Creates the rotation matrix
+        rotationMat = getRotationMatrix2D(rectangle.center, angle, 1.0);
+
+        //Perform the affine transformation
+        Imgproc.warpAffine(mat, rotatedImg, rotationMat, mat.size(), INTER_CUBIC);
+
+        //Crop the resulting image
+        getRectSubPix(rotatedImg, rect_size, rectangle.center, croppedImg);
+        return croppedImg;
+
+    }
+
+
+    /**
+     * Converts a matrix into a Bitmap and saves it in the default temp-file dir
+     * @param matrix the matrix to be converted
+     * @param imageName the name of the file being saved
+     * @author Thomas Porro(g1), Oscar Garrido(g1)
+     */
+    private void save(Mat matrix, String imageName) {
+        
+        Bitmap image = conversionMatToBitmap(matrix);
+
+        OutputStream outStream = null;
+
+        //if not specified, the system-dependent default temporary-file directory will be used
+        File tmpFile = File.createTempFile(tmpPrefix);
```

Comment:
Is this correctly? Could be useful this https://developer.android.com/training/data-storage/files#WriteCacheFileInternal ?

Reply:
Just to understand your question, you want to know if I should create a _Context_ object in which I would call the `.getCacheDir()` method, like in the example? If so, I was going to do it, but reading the Javadoc of `File.createTempFile` it should be unnecessary, since if the parameter of the directory is left blank the method will automatically pick the Cache Directory (I could better specify this in the above comment perhaps)

Reply:
I'm sorry, I've not explained well, for the cache directory you are right but anyway is the suffix parameter needed too? I've seen it here https://developer.android.com/reference/java/io/File#createTempFile(java.lang.String,%20java.lang.String) .
And I don't find where the tmpPrefix is instantiated, have I missed something?
Thanks in advance.

Reply:
Both my fault, I was so concentrated on the directory that I completely forgot the suffix. And you can't find where I initialized tmpPrefix because I didn't, sorry. I'll correct it now

Reply:
Don't worry, never mind, it's important that we make good code together.